java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/protocol/Server.java

```diff
@@ -1185,4 +1192,77 @@ private static void hackNoReuseOfOutputsForAntProjects() {
             }
         }
     }
+
+    public static class LSPServerTelemetryFactory extends CustomIndexerFactory {
+
+        private static LSPServerTelemetryFactory INSTANCE;
+        private final CustomIndexer noOp = new CustomIndexer() {
+            @Override
+            protected void index(Iterable<? extends Indexable> files, Context context) {
+            }
+        };
+        private WeakHashMap<LanguageClient, Future<Void>> clients = new WeakHashMap<>();
+
+        @MimeRegistration(mimeType="", service=CustomIndexerFactory.class)
+        public static LSPServerTelemetryFactory getDefault() {
+            if (INSTANCE == null) {
+                INSTANCE = new LSPServerTelemetryFactory();
+            }
+            return INSTANCE;
+        }
+
+        private LSPServerTelemetryFactory() {
+        }
+
+        public synchronized void connect(LanguageClient client, Future<Void> future) {
+            clients.put(client, future);
+        }
+
+        @Override
+        public synchronized boolean scanStarted(Context context) {
+            for (Map.Entry<LanguageClient, Future<Void>> entry : clients.entrySet()) {
+                if (!entry.getValue().isDone()) {
+                    entry.getKey().telemetryEvent("nbls.scanStarted");
+                }
```

Comment:
Consider removing the entry in case `entry.getValue().isDone() == true`.

Reply:
Done